Parquet parallel scan

korowa opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #137.

# Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

Improved performance for reading single parquet file / parquet files in quantity less than number of cores in multicore runtime

# What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

1) `repartition_file_scans` & `repartition_file_min_size` optimizer settings - by default repartitioning of file scans disabled, and performed if total size of files to scan greater than 10MB  (to avoid splitting small amount of small files)
2) `get_repartitioned` in ParquetExec - returns cloned object with ranged `PartitionedFile`s redistributed file groups in `base_config`
3) `repartition.rs` - now calls `get_repartitioned` for `ParquetExec` in case repartitioning is allowed (upstream operator benefit from it / no data ordering violations / etc.)

As any other repartitioning operation, parallelization is applied only in case ParquetExec is underloaded in terms of partitions -- two files scan will be distributed over 4 partitions, but no redistribution will be performed in case of "2 files - 2 target partitions".

# Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

Tests for `ParquetExec.get_repartitioned` added, and more tests added for `repartition` rule -- mostly copies of existing tests for cases when parallelization should be ignored, to ensure it won't break physical plans.

# Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

New configuration settings `repartition_file_scans` & `repartition_file_min_size`

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

tustvold commented:
Only had time to take a brief look at this PR, and so I'm likely missing something but please bear with me :smile: 

This PR modifies `ListingTable` to pair together `PartitionedFile` with `Vec<Option<FileRange>>`, this makes this approach specific to `ListingTable` and also adds parallelism control to a part of the system that doesn't really have context on how much parallelism is needed, nor what invariants such as sort orders may need to be upheld.

I have two suggestions that may be stupid:

* Make this a physical optimizer rule that looks at operators containing `FileScanConfig` and adds more partitions based on the `target_partitions` property
* Rather than adding a new `FileRanges` property, instead using the existing `range: Option<FileRange>` already present on `PartitionedFile`, the same file with disjoint ranges can then appear in multiple partitions

tustvold commented on datafusion/core/src/datasource/file_format/parquet.rs:
```diff
@@ -539,6 +542,44 @@ async fn fetch_statistics(
+    };
+
+    if collect_file_ranges {
+        let file_ranges = parquet_metadata
```
FWIW the way these ranges are applied in parquet is based on if the row group's midpoint lies within the given range, as a result there is no requirement that these ranges exactly delimit boundaries.

For example you could take a parquet file of 2GB and blindly chop it into 4x 512MB slices. This makes the assumption that there are at least 4 row groups and the row groups are similarly sized, in practice this is probably fine. This is what Spark does

korowa replied:
True, cutting file on N even parts will allow to read only row groups having their start offset inside corresponding ranges without any duplicate reads or skipped row groups - so, splitting could be much easier without using metadata (except for ObjectMetadata for the size)

korowa commented:
@tustvold , thank you for the comments!

Initially my intention was to handle scan planning as early as possible, so `ListingTable` looked like proper place for this - it actually holds parallelism settings in `options` attribute, and fetches required metadata.

But, yeah, now I see that physical optimizer, and especially its `repartitioning` rule is much better suited for repatitioning ParquetExec 🤔 .

I guess I'll convert this PR to draft and come up a bit later with updated version of this optimizer rule.

korowa commented:
I've reworked this PR by utilizing `repartition` rule, and updated the description -- it looks way more simple now, and, what's more important, it takes into account any plan-related restrictions for ParquetExec.

@tustvold, thank you for the tip about using physical optimizer!

tustvold commented on datafusion/core/src/physical_plan/file_format/parquet.rs:
```diff
@@ -232,6 +241,74 @@ impl ParquetExec {
+            .flatten()
+            .map(|f| f.object_meta.size as i64)
+            .sum::<i64>();
+        let target_partition_size =
```
I think we should probably have a minimum file range, anything less than 10MB is unlikely to yield additional parallelism, as row groups shouldn't be this small, and adding more partitions is not necessarily free.

We could make this configurable

korowa replied:
Sounds reasonable, but it seems to me that same thing can be done via both `target_partitions` & `parallel_file_scan` if we know file layout prior to running query - is this setting supposed to be kind of fuse for running queries over random files (i.e. neither the number of files nor their size is known)?

korowa replied:
After some thoughts, I see that it could be just more user-friendly -- cause this setting will care about preventing unnecessary file splits without any additional settings tweaking efforts.

I've also moved both of these new settings from `execution` to `optimizer` section.

tustvold reviewed (approved):
This looks really cool, left some minor comments

korowa commented on datafusion/core/src/physical_optimizer/repartition.rs:
```diff
@@ -686,24 +729,6 @@ mod tests {
         Ok(())
     }
 
```
By accident, now it's back

alamb commented:
I plan to review this carefully either later today or tomorrow. Very exciting @korowa  -- thank you

alamb reviewed (approved):
I went through the code carefully and I really like it. Thank you @korowa -- do you have any performance benchmarks you can share? I think this will mostly help when scanning single large parquet files. 

I would like to explore turning this feature on by default (perhaps we can have a separate ticket to track that)

I already feel bad that we don't have other parquet options enabled by default.

alamb commented:
My measurements suggest this setting can improve the performance with single large parquet files significantly (over 2x in my measurement). 👨‍🍳 👌   -- very nice

I tested this out by making a 9G parquet file from https://github.com/tustvold/access-log-gen/

Then using datafusion-cli:

```sql
❯ select avg(request_bytes), avg(response_bytes), avg(response_status), host from '/Users/alamb/Software/access-log-gen/logs.9G.parquet' group by host;
...
927 rows in set. Query took 2.313 seconds.
```

And then I enabled this setting:

```sql
❯ set datafusion.optimizer.repartition_file_scans = true;
0 rows in set. Query took 0.000 seconds.
❯ select avg(request_bytes), avg(response_bytes), avg(response_status), host from '/Users/alamb/Software/access-log-gen/logs.9G.parquet' group by host;

927 rows in set. Query took 0.962 seconds.
```

😮

korowa commented:
I have only these in my notes
```
// Single partition
./target/release/examples/clickbench  4.10s user 0.59s system 113% cpu 4.142 total
./target/release/examples/clickbench  4.14s user 0.56s system 136% cpu 3.451 total
./target/release/examples/clickbench  4.13s user 0.55s system 135% cpu 3.446 total
./target/release/examples/clickbench  4.29s user 0.55s system 134% cpu 3.603 total

// Multiple partition
./target/release/examples/clickbench  5.32s user 0.96s system 288% cpu 2.174 total
./target/release/examples/clickbench  5.14s user 0.89s system 392% cpu 1.534 total
./target/release/examples/clickbench  5.24s user 0.92s system 405% cpu 1.517 total
./target/release/examples/clickbench  5.14s user 0.91s system 399% cpu 1.516 total
./target/release/examples/clickbench  5.23s user 0.92s system 405% cpu 1.518 total
```
which doesn't look like "official" benchmark at all. These are results for [this query from clickbench](https://github.com/ClickHouse/ClickBench/blob/main/datafusion/queries.sql#L26) over 14GB parquet file on 2.6 GHz 6-Core Intel Core i7

korowa commented on datafusion/core/src/physical_optimizer/repartition.rs:
```diff
@@ -846,6 +897,182 @@ mod tests {
+        Ok(())
+    }
+
+    #[test]
+    fn parallelization_two_partitions() -> Result<()> {
+        let plan = aggregate(parquet_exec_two_partitions());
+
+        let expected = [
+            "AggregateExec: mode=Final, gby=[], aggr=[]",
+            "CoalescePartitionsExec",
+            "AggregateExec: mode=Partial, gby=[], aggr=[]",
+            // Plan already has two partitions
+            "ParquetExec: limit=None, partitions={2 groups: [[x], [y]]}, projection=[c1]",
+        ];
+
+        assert_optimized!(expected, plan, 2, true, 10);
+        Ok(())
+    }
```
Good one, done.

korowa commented:
> I would like to explore turning this feature on by default (perhaps we can have a separate ticket to track that)

I don't mind enabling parallelism by default and it seems to be the fastest way to deliver this feature, but (I'm not sure, just a suggestion) maybe better time for this will be in 1 (or 2) releases after the setting itself will be released?

alamb commented:
> As it works for now - yes, use case is mostly "relatively large files less than number of target_partitions" -- I guess it could be improved / reworked later to something like "perform repartitioning even for target_partitions in case there is significant skew in current partitioning"

I think to make this really happen we will need to have more runtime dynamics (aka using a morsel driven sceheduler)

> I don't mind enabling parallelism by default and it seems to be the fastest way to deliver this feature, but (I'm not sure, just a suggestion) maybe better time for this will be in 1 (or 2) releases after the setting itself will be released?

I agree -- let's get this PR merged in (default to off) and then plan to enable it by default in a few weeks (we just need to remember to do so!)

alamb commented on datafusion/core/src/physical_optimizer/repartition.rs:
```diff
@@ -928,6 +928,22 @@ mod tests {
+            "CoalescePartitionsExec",
+            "AggregateExec: mode=Partial, gby=[], aggr=[]",
+            // Multiple source files splitted across partitions
+            "ParquetExec: limit=None, partitions={4 groups: [[x:0..75], [x:75..100, y:0..50], [y:50..125], [y:125..200]]}, projection=[c1]",
```
that is quite clever that the partitions have different parts of the same file 👍

alamb commented:
I plan to leave this open for the rest of the weekend so others have a chance to comment if they want, and then merge on Monday

alamb commented:
Thanks again @korowa ❤️

Ted-Jiang commented:
Thanks @korowa  this is really cool!  👍  nice work!
I have a question about `multiple row groups from the same file may be read concurrently`

from the code 
```
 let target_partition_size =
            (total_size as usize + (target_partitions) - 1) / (target_partitions);
```
How could we make sure the file is dived into row group? 🤔

alamb commented:
I filed https://github.com/apache/arrow-datafusion/issues/5125 to track turning this on by default

tustvold commented:
> How could we make sure the file is dived into row group?

Without the parquet file metadata we can't reliably, but it isn't important to correctness that we do so. Not needing the metadata significantly simplifies the planning and avoids potentially costly round trips to object store whilst planning. FWIW I believe this is a similar approach as taken by Spark.

alamb commented:
I believe what happens is that the file is divided into byte ranges and then the row groups whose data falls within those ranges are scanned. It would be good to double check this undertanding though

https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/datasource/listing/mod.rs#L52-L63

tustvold commented:
> whose data falls within those ranges are scanned

More specifically it is the row groups with a midpoint that falls into the range, this means that so long as the ranges are disjoint there is no risk of reading the same row group twice

korowa commented:
@mingmwang here is the [exact place](https://github.com/apache/arrow-datafusion/blob/125a8580c19c78c99fbbe3a6afe373de2538b205/datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs#L57) where DF decides to read/not too read RowGroup depending on range.

Ted-Jiang commented:
Thanks for all kindly reply !   ❤️
 > here is the [exact place](https://github.com/apache/arrow-datafusion/blob/125a8580c19c78c99fbbe3a6afe373de2538b205/datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs#L57) where DF decides to read/not to read RowGroup depending on range. So it actually isn't required to split files on ranges with boundaries same as RowGroups boundaries.

So `PartitionedFile ` range is not used for fetch bytes from objectStore, finally we use rowgroup start, offset to fetch bytest ?  am i right?

korowa commented:
> So `PartitionedFile ` range is not used for fetch bytes from objectStore, finally we use rowgroup start, offset to fetch bytest ? am i right?

Yeah, I can't find any places where range is used, except for row_group pruning

Ted-Jiang commented:
> > So `PartitionedFile ` range is not used for fetch bytes from objectStore, finally we use rowgroup start, offset to fetch bytest ? am i right?
> 
> Yeah, I can't find any places where range is used, except for row_group pruning

Thanks again for everyone ✌️
